@xen-orchestra/audit-core/src/index.js

```diff
@@ -0,0 +1,110 @@
+// see https://github.com/babel/babel/issues/8450
+import 'core-js/features/symbol/async-iterator'
+
+import assert from 'assert'
+import defer from 'golike-defer'
+import hash from 'object-hash'
+
+export class Storage {
+  constructor() {
+    this._lock = Promise.resolve()
+  }
+
+  async acquireLock() {
+    const lock = this._lock
+    let releaseLock
+    this._lock = new Promise(resolve => {
+      releaseLock = resolve
+    })
+    await lock
+    return releaseLock
+  }
+}
+
+// Format: $<algorithm>$<salt>$<encrypted>
+//
+// http://man7.org/linux/man-pages/man3/crypt.3.html#NOTES
+const ID_TO_ALGORITHM = {
+  '5': 'sha256',
+}
+
+export const NULL_ID = 'nullId'
+
+const HASH_ALGORITHM_ID = '5'
+const createHash = (data, algorithmId = HASH_ALGORITHM_ID) =>
+  `$${algorithmId}$$${hash(data, {
+    algorithm: ID_TO_ALGORITHM[algorithmId],
+    excludeKeys: key => key === 'id',
+  })}`
+
+export class AuditCore {
+  constructor(storage) {
+    assert(storage !== undefined)
+    this._storage = storage
+  }
+
+  @defer
+  async add($defer, subject, event, data) {
+    const time = Date.now()
+    const storage = this._storage
+    $defer(await storage.acquireLock())
+    const record = {
+      data,
+      event,
+      previousId: (await storage.getLastId()) ?? NULL_ID,
+      subject,
+      time,
+    }
+    record.id = createHash(record)
+    await storage.put(record)
+    await storage.setLastId(record.id)
+    return record
+  }
+
+  async _getOldestValidatedId(oldest, newest) {
+    while (newest !== oldest) {
+      const record = await this._storage.get(newest)
+      if (record === undefined) {
+        return {
+          id: newest,
+          reason: `unable to reach the record ${oldest} (stopped at ${newest})`,
+        }
+      }
+      if (
+        newest !== createHash(record, newest.slice(1, newest.indexOf('$', 1)))
+      ) {
+        return {
+          id: newest,
```

Comment:
Incorrect!

Reply:
IMHO, we stopped at newest.

Reply:
My bad it's correct: we return the identifier of the first non-validated record.